presto-spark-base/src/main/java/com/facebook/presto/spark/execution/AbstractPrestoSparkQueryExecution.java

```diff
@@ -708,6 +717,144 @@ else if (writerTarget instanceof ExecutionWriterTarget.RefreshMaterializedViewHa
+            ImmutableList.Builder<PrestoSparkBroadcastDependency<?>> broadcastDependencies = ImmutableList.builder();
+
+            for (SubPlan child : subPlan.getChildren()) {
+                RddAndMore<?> childRdd = createRddFragmented(child, tableWriteInfo);
+                if (childRdd.getPartitioningHandle().isPresent() && childRdd.getPartitioningHandle().get().equals(FIXED_BROADCAST_DISTRIBUTION)) {
+                    PrestoSparkBroadcastDependency<?> broadcastDependency = createBroadcastDependency(childRdd);
+                    broadcastInputs.put(child.getFragment().getId(), broadcastDependency.executeBroadcast(sparkContext));
+                    broadcastDependencies.add(broadcastDependency);
+                }
+                else {
+                    rddInputs.put(child.getFragment().getId(), (JavaPairRDD<MutablePartitionId, PrestoSparkMutableRow>) childRdd.getRdd());
+                    broadcastDependencies.addAll(childRdd.getBroadcastDependencies());
+                }
+            }
+
+            Class outputType = PrestoSparkSerializedPage.class;
+            if (!isRootSubPlan(subPlan)) {
+                if (isBroadcastDistribution(subPlan) && isStorageBasedBroadcastJoinEnabled(session)) {
+                    outputType = PrestoSparkStorageHandle.class;
+                }
+                else {
+                    outputType = PrestoSparkMutableRow.class;
+                }
+            }
+
+            JavaPairRDD<MutablePartitionId, T> rdd = rddFactory.createSparkRdd(
+                    sparkContext,
+                    session,
+                    subPlan.getFragment(),
+                    rddInputs.build(),
+                    broadcastInputs.build(),
+                    taskExecutorFactoryProvider,
+                    taskInfoCollector,
+                    shuffleStatsCollector,
+                    tableWriteInfo,
+                    outputType);
+
+            if (subPlan.getFragment().getId().getId() != 0 && !isBroadcastDistribution(subPlan)) {
+                rdd = (JavaPairRDD<MutablePartitionId, T>) partitionBy(subPlan.getFragment().getId().getId(), (JavaPairRDD<MutablePartitionId, PrestoSparkMutableRow>) rdd, subPlan.getFragment().getPartitioningScheme());
+            }
+
+            RddAndMore rddAndMore = new RddAndMore<T>(rdd, broadcastDependencies.build(), Optional.ofNullable(subPlan.getFragment().getPartitioningScheme().getPartitioning().getHandle()));
+            fragmentIdToRdd.put(subPlan.getFragment().getId(), rddAndMore);
+            return rddAndMore;
+        }
+    }
+
+    private boolean isBroadcastDistribution(SubPlan subPlan)
+    {
+        return subPlan.getFragment().getPartitioningScheme().getPartitioning().getHandle().equals(FIXED_BROADCAST_DISTRIBUTION);
+    }
+
+    private boolean isRootSubPlan(SubPlan subPlan)
+    {
+        return subPlan.getFragment().getId().getId() == 0;
```

Comment:
Let's not hardcode 0 here. I think Rebecca had added a constant for the root id.